addons/web_editor/static/lib/odoo-editor/test/spec/format.test.js

```diff
@@ -961,6 +961,13 @@ describe('setTagName', () => {
                 contentAfter: '<h1 class="text-uppercase">[abcd]</h1>',
             });
         });
+        it('should turn a inline elements within a div into a heading 1', async () => {
+            await testEditor(BasicEditor, {
+                contentBefore: '<div><font style="color: red;">a[b</font><span style="font-size: 12px;">c]d</span><font style="color: red;">ef</font><p>gh</p><font style="color: red;">ij</font><span style="font-size: 12px;">kl</span></div>',
+                stepFunction: editor => editor.execCommand('setTag', 'h1'),
+                contentAfter: '<div><h1><font style="color: red;">a[b</font><span style="font-size: 12px;">c]d</span><font style="color: red;">ef</font></h1><p>gh</p><font style="color: red;">ij</font><span style="font-size: 12px;">kl</span></div>',
```

Comment:
I'm a bit unclear on whether we should handle this case in the first place so I have doubts on this fix. I agree that `<div>a<div>` -> `setTag('h1')` should lead to `<div><h1>a</h1></div>`. But then we have no way to get back to `<div>a</div>` unless we decide that `setTag('p')` should understand that we're in a case where the `div` is sufficient. But how can we possibly know? Once we have cases like `<div><h5>[a]</h5>bc<p>d</p></div>` -> `setTag('p')`, we're completely stuck as nothing is gonna be right. Nothing we can do will be satisfactory. IIRC, this is already the reason we currently force `<p>`s in `<td>`s. Similar questions occur when pressing ENTER, for instance between "b" and "c".
So I'd argue we shouldn't support a DOM like this and any template that does this (eg, "big icons subtitles") should be adapted. But that's a big move. What do you think @qsm-odoo @rdeodoo @pierre-pde @Goaman @sebgeelen ?

While we discuss this, it may be wise to separate this fix from the rest of the PR so we can finish and merge the rest soon.

Reply:
> So I'd argue we shouldn't support a DOM like this and any template that does this (eg, "big icons subtitles") should be adapted.

I suggested the same to @dmo-odoo he replied with:

> DMO's reply: Unfortunately, you can't change the xml of snippets in stable. Anyway, I think that setTag should be fixed to recognize cases like that.

For context current working of a structure like is
`<div><span>ab[c</span><font>de]f</font><h1>abc</h1><span>abc</span><font>def</font></div>` => `setTag('h2')` => `<div><h2><span>ab[c</span><font>de]f</font><h1>abc</h1><span>abc</span><font>def</font></h2></div>` which is completely incorrect what I have done is it wraps all consecutive inline nodes to the tag selected, here according to my fix it would be `<div><h2><span>ab[c</span><font>de]f</font></h2><h1>abc</h1><span>abc</span><font>def</font></div>`. But I understand that converting it back to `normal`, would be confusing on whether to unwrap the `h2` or convert `h2` to `p`

Also I have created another PR https://github.com/odoo/odoo/pull/145734 to continue the discussion...

Reply:
> For context current working for a DOM like this is
<div><span>ab[c</span><font>de]f</font><h1>abc</h1><span>abc</span><font>def</font></div> => setTag('h2') => <div><h2><span>ab[c</span><font>de]f</font><h1>abc</h1><span>abc</span><font>def</font></h2></div> which is completely incorrect

Yes indeed if we indeed allow this content, the current behavior is incorrect. The question to me is whether we should fix the behavior or fix the original HTML. David is correct that we can't change it in stable, but if we decide the HTML is incorrect we can always detect it and correct it automatically with js. Not great, but I'm not sure we can do much better.